File: Commercial/Commercial_ControllerGo/Battery.go
package main

import (
	"fmt"
	"os"
)

// STATIC PROPERTIES (in Golang, it's treated as a global)
var NumFloors int
var NumBasements int
var NumColumns int

type Battery struct {

	// PROPERTIES
	ID                  int
	Status              string // online|offline
	ColumnList          []Column
	BoardButtonList     []BoardButton
	isFire              bool
	isPowerOutage       bool
	isMechanicalFailure bool
}

// CONSTRUCTOR :-> Initialize a column with the default values when creating a battery's columns list
func (battery *Battery) InitBattery(id int) {

	battery.ID = id
	battery.Status = "online" // online|offline
	battery.ColumnList = []Column{}
	battery.BoardButtonList = []BoardButton{}
	battery.isFire = false
	battery.isPowerOutage = false
	battery.isMechanicalFailure = false
}

// METHODS
// Run all initialiazing/startup Battery methods
func (battery *Battery) Run() {

	battery.CreateColumnList()
	battery.CreateBoardButtons()
}

// Initialize the battery's collection of columns
func (battery *Battery) CreateColumnList() {

	for columnID := 1; columnID <= NumColumns; columnID++ {

		column := Column{}
		column.InitColumn(columnID)

		// Set up allowed floor ranges
		if NumBasements > 0 {

			if columnID == 1 {

				// Column takes care of basement floors
				column.LowestFloor = -(NumBasements)
				column.HighestFloor = -1

			} else {

				// Column takes care of above-ground floors
				column.LowestFloor = 1 + NumFloors/(NumColumns-1)*(columnID-2)
				column.HighestFloor = NumFloors / (NumColumns - 1) * (columnID - 1)

			}

		} else {

			// No basement floors - therefore all floors are above-ground
			column.LowestFloor = 1 + NumFloors/NumColumns*(columnID-1)
			column.HighestFloor = NumFloors / NumColumns * columnID
		}

		column.CreateElevatorList()
		column.CreateCallButtons()
		battery.ColumnList = append(battery.ColumnList, column)
	}
}

// Initialize the battery's board display's buttons
func (battery *Battery) CreateBoardButtons() {

	// Board buttons for basements floors
	for numBasement := -(NumBasements); numBasement < 0; numBasement++ {

		button := BoardButton{}
		button.InitBoardButton(numBasement, battery)
		battery.BoardButtonList = append(battery.BoardButtonList, button)
	}

	// Board buttons for non-basement floors
	for numFloor := OriginFloor; numFloor <= NumFloors; numFloor++ {

		button := BoardButton{}
		button.InitBoardButton(numFloor, battery)
		battery.BoardButtonList = append(battery.BoardButtonList, button)
	}
}

// Monitor the battery's elevator system
// In real conditions, the monitoring would be done in a near-infinite while loop
func (battery *Battery) MonitorSystem() {

	shouldExit := false

	if battery.isFire || battery.isPowerOutage || battery.isMechanicalFailure {

		battery.Status = "offline"

		for i := range battery.ColumnList {

			battery.ColumnList[i].Status = "offline"
			for j := range battery.ColumnList[i].ElevatorList {

				battery.ColumnList[i].ElevatorList[j].Status = "offline"
			}
		}
		fmt.Printf("Battery %d has been shut down for maintenance. Sorry for the inconvenience.", battery.ID)
		shouldExit = true
	}

	// Stop execution of script if something is wrong
	if shouldExit {
		os.Exit(-1)
	}
}

// Assigns a column and elevator to use when making a request from RC via the board buttons
func (battery *Battery) AssignElevator(requestedFloor int) {

	// Select the call button to press
	var boardBtnToPress BoardButton
	for i := range battery.BoardButtonList {

		if battery.BoardButtonList[i].RequestedFloor == requestedFloor {
			boardBtnToPress = battery.BoardButtonList[i]
			break
		}
	}

	// Select the chosen elevator
	var chosenElevator Elevator
	if elevator, err := boardBtnToPress.Press(); err == nil {
		chosenElevator = elevator

	} else {
		fmt.Println("AssignElevator(), Battery.go FAILED: ", err)
	}

	// Do requests until elevator has reached the floor where the call was made (RC floor)
	for chosenElevator.CurrentFloor != OriginFloor {
		chosenElevator.DoRequests()
	}

	// Set a request for the elevator to go to requested floor, once picked up
	newDirection := ""
	if chosenElevator.CurrentFloor < requestedFloor {
		newDirection = "up"

	} else {
		newDirection = "down"
	}

	chosenElevator.SendRequest(requestedFloor, newDirection)

	// Do requests until elevator has reached requested floor
	for chosenElevator.CurrentFloor != requestedFloor {
		chosenElevator.DoRequests()
	}
}

File: Commercial/Commercial_ControllerCS/Commercial_ControllerCS/Battery.cs
using System;
using System.Collections.Generic;
using static System.Console;

namespace Commercial_ControllerCS
{
    class Battery
    {
        #region STATIC FIELDS
        private static int _numFloors;
        private static int _numBasements;
        private static int _numColumns;
        #endregion

        #region STATIC PROPERTIES
        public static int NumFloors {
            get { return _numFloors; }
            set
            {
                if (value <= 0)
                    throw new Exception("There has to be more than 0 floors, excluding the basements.");
                else
                    _numFloors = value;
            }
        }
        public static int NumBasements
        {
            get { return _numBasements; }
            set
            {
                if (value < 0)
                    throw new Exception("There has to be 0 or more basements.");
                else
                    _numBasements = value;
            }
        }
        public static int NumColumns
        {
            get { return _numColumns; }
            set
            {
                if (value <= 0)
                    throw new Exception("There has to be more than 0 columns.");
                else
                    _numColumns = value;
            }
        }

        #endregion

        #region FIELDS
        private string _status;
        #endregion

        #region PROPERTIES
        public int ID { get; private set; }
        public string Status
        {
            get { return _status; }
            set
            {
                if (value.ToLower() != "online" && value.ToLower() != "offline")
                    throw new Exception($"Invalid value for battery {ID}'s status. Can only be either 'online' or 'offline'.");
                else
                    _status = value;
            }
        }
        public bool isFire { get; set; }
        public bool isPowerOutage { get; set; }
        public bool isMechanicalFailure { get; set; }
        public List<Column> ColumnList { get; set; }
        public List<BoardButton> BoardButtonList { get; set; }
        #endregion

        #region CONSTRUCTOR
        public Battery(int id_)
        {
            ID = id_;
            Status = "online"; // online|offline
            ColumnList = new List<Column>();
            BoardButtonList = new List<BoardButton>();
            isFire = false;
            isPowerOutage = false;
            isMechanicalFailure = false;
        }
        #endregion

        #region STATIC METHODS
        // Quickly change the Battery's static properties
        public static void ChangeProperties(int numColumns_, int numFloors_, int numBasements_)
        {
            NumColumns = numColumns_;
            NumFloors = numFloors_;
            NumBasements = numBasements_;
        }
        #endregion

        #region METHODS
        // Run all initialiazing/startup Battery methods
        public void Run()
        {
            CreateColumnList();
            CreateBoardButtons();
        }

        // Initialize the battery's collection of columns
        public void CreateColumnList()
        {
            for (int columnID = 1; columnID <= NumColumns; columnID++)
            {
                var column = new Column(columnID);
      
                // Set up allowed floor ranges
                if (NumBasements > 0)
                {
                    if (columnID == 1)
                    {
                        // Column takes care of basement floors
                        column.LowestFloor = -(NumBasements);
                        column.HighestFloor = -1;
                    }
                    else
                    {
                        // Column takes care of above-ground floors
                        column.LowestFloor = 1 + NumFloors / (NumColumns - 1) * (columnID - 2);
                        column.HighestFloor = NumFloors / (NumColumns - 1) * (columnID - 1);
                    }
                }
                else
                {
                    // No basement floors - therefore all floors are above-ground
                    column.LowestFloor = 1 + NumFloors / NumColumns * (columnID - 1);
                    column.HighestFloor = NumFloors / NumColumns * columnID;
                }

                column.CreateElevatorList();
                column.CreateCallButtons();
                ColumnList.Add(column);
            }
        }

        // Initialize the battery's board display's buttons
        public void CreateBoardButtons()
        {
            // Board buttons for basements floors
            for (int numBasement = -(NumBasements); numBasement < 0; numBasement++)
            {
                var button = new BoardButton(numBasement, this);
                BoardButtonList.Add(button);
            }

            // Board buttons for non-basement floors
            for (int numFloor = Elevator.OriginFloor; numFloor <= NumFloors; numFloor++)
            {
                var button = new BoardButton(numFloor, this);
                BoardButtonList.Add(button);
            }

        }

        // Monitor the battery's elevator system
        // In real conditions, the monitoring would be done in a near-infinite while loop
        public void MonitorSystem()
        {
            if (isFire || isPowerOutage || isMechanicalFailure)
            {
                Status = "offline";
                foreach (var column in ColumnList)
                {
                    column.Status = "offline";
                    foreach (var elevator in column.ElevatorList)
                    {
                        elevator.Status = "offline";
                    }
                }

                // Stop execution of script
                Write($"Battery {ID} has been shut down for maintenance. Sorry for the inconvenience.");
                System.Environment.Exit(1);
            }
        }

        // Assigns a column and elevator to use when making a request from RC via the board buttons
        public void AssignElevator(int requestedFloor)
        {
            var boardBtnToPress = BoardButtonList.Find(btn => btn.RequestedFloor == requestedFloor);
            Elevator chosenElevator = boardBtnToPress.Press();

            // Do requests until elevator has reached the floor where the call was made (RC floor)
            while (chosenElevator.CurrentFloor != Elevator.OriginFloor)
                chosenElevator.DoRequests();

            // Set a request for the elevator to go to requested floor, once picked up
            string newDirection = (chosenElevator.CurrentFloor < requestedFloor) ? "up" : "down";
            var newRequest = new Request(requestedFloor, newDirection);
            chosenElevator.RequestsQueue.Add(newRequest);
           

            // Do requests until elevator has reached requested floor 
            while (chosenElevator.CurrentFloor != requestedFloor)
                chosenElevator.DoRequests();

        }
        #endregion
    }
}
